pkg/controller/lookup_cache.go

```diff
@@ -72,8 +72,9 @@ func (c *MatchingCache) Add(labelObj objectWithMeta, selectorObj objectWithMeta)
 // we need check in the external request to ensure the cache data is not dirty.
 func (c *MatchingCache) GetMatchingObject(labelObj objectWithMeta) (controller interface{}, exists bool) {
 	key := keyFunc(labelObj)
-	c.mutex.RLock()
-	defer c.mutex.RUnlock()
+	// NOTE: we use Lock() instead of RLock() here because lru's Get() method is not threadsafe
```

Comment:
Get() method modifies state? If it does not modify the state, RLock is fine. We do not need it to be threadsafe. That is the whole point of the lock here.